File: _includes/code/java-v6/src/test/java/_StarterGuidesCustomVectorsTest.java
// import com.fasterxml.jackson.annotation.JsonProperty;
// import com.fasterxml.jackson.core.type.TypeReference;
// import com.fasterxml.jackson.databind.ObjectMapper;
// import io.weaviate.client6.v1.api.WeaviateClient;
// import io.weaviate.client6.v1.api.collections.CollectionHandle;
// import io.weaviate.client6.v1.api.collections.VectorConfig;
// import io.weaviate.client6.v1.api.collections.WeaviateObject;
// import io.weaviate.client6.v1.api.collections.query.Metadata;
// import org.junit.jupiter.api.Test;

// import java.net.URI;
// import java.net.http.HttpClient;
// import java.net.http.HttpRequest;
// import java.net.http.HttpResponse;
// import java.util.ArrayList;
// import java.util.List;
// import java.util.Map;

// import static org.assertj.core.api.Assertions.assertThat;

// class StarterGuidesCustomVectorsTest {

//   // Helper class for parsing the JSON data with vectors
//   private static class JeopardyQuestionWithVector {
//     @JsonProperty("Answer")
//     String answer;
//     @JsonProperty("Question")
//     String question;
//     @JsonProperty("Category")
//     String category;
//     @JsonProperty("vector")
//     float[] vector;
//   }

//   @Test
//   void testBringYourOwnVectors() throws Exception {
//     WeaviateClient client = null;
//     String collectionName = "Question";

//     try {
//       // Clean slate
//       client = WeaviateClient.connectToLocal();
//       if (client.collections.exists(collectionName)) {
//         client.collections.delete(collectionName);
//       }

//       // ===== Create schema =====
//       // Create the collection.
//       client.collections.create(collectionName,
//           col -> col.vectorConfig(VectorConfig.selfProvided()));

//       // ===== Import data =====
//       String fname = "jeopardy_tiny_with_vectors_all-OpenAI-ada-002.json";
//       String url =
//           "https://raw.githubusercontent.com/weaviate-tutorials/quickstart/main/data/" + fname;

//       HttpClient httpClient = HttpClient.newHttpClient();
//       HttpRequest request = HttpRequest.newBuilder().uri(URI.create(url)).build();
//       HttpResponse<String> responseHttp =
//           httpClient.send(request, HttpResponse.BodyHandlers.ofString());
//       String responseBody = responseHttp.body();

//       ObjectMapper objectMapper = new ObjectMapper();
//       List<JeopardyQuestionWithVector> data =
//           objectMapper.readValue(responseBody, new TypeReference<>() {});

//       CollectionHandle<Map<String, Object>> questions = client.collections.use(collectionName);
//       List<WeaviateObject> questionObjs = new ArrayList<>();

//       for (JeopardyQuestionWithVector d : data) {
//         // highlight-start
//         Map<String, Object> properties = Map.of(
//             "answer", d.answer,
//             "question", d.question,
//             "category", d.category
//         );
//         questionObjs.add(WeaviateObject.of(
//             properties,
//             // This is the "instruction" function (lambda) that adds the vector
//             obj -> obj.vector(Vectors.of("title_vec", d.vector))
//         ));
//         // highlight-end
//       }

//       CollectionHandle<Map<String, Object>> questions = client.collections.use(collectionName);
//       questions.data.insertMany(questionObjs.toArray(new WeaviateObject[0]));

//       // ===== Query =====
//       float[] queryVector = new float[] {0.0042927247f, -0.007413445f, 0.00034457954f,
//           /* ... shortened for brevity ... */ -0.025992135f};

//       // Added a small sleep to ensure indexing is complete
//       Thread.sleep(2000);

//       var response = questions.query.nearVector(queryVector,
//           q -> q.limit(2).returnMetadata(Metadata.CERTAINTY));

//       System.out.println(response);

//       // ===== Test query results =====
//       assertThat(response.objects()).hasSize(2);
//       assertThat(response.objects().get(0).metadata().certainty()).isNotNull()
//           .isInstanceOf(Double.class);
//       assertThat(response.objects().get(0).properties()).isNotNull().isInstanceOf(Map.class);

//     } finally {
//       if (client != null) {
//         if (client.collections.exists(collectionName)) {
//           client.collections.delete(collectionName);
//         }
//         client.close();
//       }
//     }
//   }
// }

File: _includes/code/csharp/StarterGuidesCustomVectorsTest.cs
using Xunit;
using Weaviate.Client;
using Weaviate.Client.Models;
using System;
using System.Threading.Tasks;
using System.Text.Json;
using System.Linq;
using System.Collections.Generic;
using System.Net.Http;
using System.Text.Json.Serialization;

namespace WeaviateProject.Tests;

public class StarterGuidesCustomVectorsTest
{
    // Helper class for parsing the JSON data with vectors
    private record JeopardyQuestionWithVector
    {
        [JsonPropertyName("Answer")]
        public string Answer { get; init; }
        [JsonPropertyName("Question")]
        public string Question { get; init; }
        [JsonPropertyName("Category")]
        public string Category { get; init; }
        [JsonPropertyName("vector")]
        public float[] Vector { get; init; }
    }

    [Fact]
    public async Task TestBringYourOwnVectors()
    {
        using var client = await Connect.Local();
        string collectionName = "Question";

        try
        {
            // Clean slate
            if (await client.Collections.Exists(collectionName))
            {
                await client.Collections.Delete(collectionName);
            }

            // START CreateCollection
            // Create the collection.
            await client.Collections.Create(new CollectionConfig
            {
                Name = collectionName,
                Properties =
                [
                    Property.Text("answer"),
                    Property.Text("question"),
                    Property.Text("category")
                ],
                // Configure the "default" vector to be SelfProvided (BYOV)
                VectorConfig = Configure.Vectors.SelfProvided().New("default")
            });
            // END CreateCollection

            // START ImportData
            var fname = "jeopardy_tiny_with_vectors_all-OpenAI-ada-002.json";
            var url = $"https://raw.githubusercontent.com/weaviate-tutorials/quickstart/main/data/{fname}";

            using var httpClient = new HttpClient();
            var responseBody = await httpClient.GetStringAsync(url);

            var data = JsonSerializer.Deserialize<List<JeopardyQuestionWithVector>>(responseBody);

            // Get a handle to the collection
            var questions = client.Collections.Use(collectionName);

            // Using Insert in a loop allows explicit vector assignment per object.
            // Using Task.WhenAll creates a parallel batch-like effect.
            var insertTasks = data.Select(d =>
            {
                // highlight-start
                return questions.Data.Insert(
                    // Pass properties as an Anonymous Type
                    data: new
                    {
                        answer = d.Answer,
                        question = d.Question,
                        category = d.Category
                    },
                    // Explicitly pass the vector
                    vectors: d.Vector
                );
                // highlight-end
            });

            await Task.WhenAll(insertTasks);
            // END ImportData

            // START NearVector
            var queryVector = data[0].Vector; // Use a vector from the dataset for a reliable query

            // Added a small delay to ensure indexing is complete
            await Task.Delay(2000);

            var response = await questions.Query.NearVector(
                queryVector,
                limit: 2,
                returnMetadata: MetadataOptions.Certainty
            );

            Console.WriteLine(JsonSerializer.Serialize(response.Objects));
            // END NearVector

            // ===== Test query results =====
            Assert.Equal(2, response.Objects.Count());
            // The first result should be the object we used for the query, with near-perfect certainty
            Assert.NotNull(response.Objects.First().Metadata.Certainty);
            Assert.True(response.Objects.First().Metadata.Certainty > 0.999);

            var props = response.Objects.First().Properties;
            Assert.Equal(data[0].Question, props["question"].ToString());
        }
        finally
        {
            if (client != null && await client.Collections.Exists(collectionName))
            {
                await client.Collections.Delete(collectionName);
            }
        }
    }
}